pkg/3scale/amp/component/backend_options.go

```diff
@@ -0,0 +1,98 @@
+	"github.com/go-playground/validator/v10"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+type BackendOptions struct {
+	ServiceEndpoint              string `validate:"required"`
+	RouteEndpoint                string `validate:"required"`
+	StorageURL                   string `validate:"required"`
+	QueuesURL                    string `validate:"required"`
+	StorageSentinelHosts         string
+	StorageSentinelRole          string
+	QueuesSentinelHosts          string
+	QueuesSentinelRole           string
+	ListenerResourceRequirements v1.ResourceRequirements `validate:"-"`
+	WorkerResourceRequirements   v1.ResourceRequirements `validate:"-"`
+	CronResourceRequirements     v1.ResourceRequirements `validate:"-"`
+	ListenerReplicas             int32
+	WorkerReplicas               int32
+	CronReplicas                 int32
+	AppLabel                     string `validate:"required"`
+	SystemBackendUsername        string `validate:"required"`
+	SystemBackendPassword        string `validate:"required"`
+	TenantName                   string `validate:"required"`
+	WildcardDomain               string `validate:"required"`
+}
+
+func NewBackendOptions() *BackendOptions {
+	return &BackendOptions{}
+}
+
+func (b *BackendOptions) Validate() error {
+	validate := validator.New()
+	return validate.Struct(b)
+}
+
+func DefaultBackendServiceEndpoint() string {
+	return "http://backend-listener:3000"
+}
+
+func DefaultBackendRedisStorageURL() string {
+	return "redis://backend-redis:6379/0"
+}
+
+func DefaultBackendRedisQueuesURL() string {
+	return "redis://backend-redis:6379/1"
+}
+
+func DefaultBackendListenerResourceRequirements() v1.ResourceRequirements {
+	return v1.ResourceRequirements{
+		Limits: v1.ResourceList{
+			v1.ResourceCPU:    resource.MustParse("1000m"),
+			v1.ResourceMemory: resource.MustParse("700Mi"),
+		},
+		Requests: v1.ResourceList{
+			v1.ResourceCPU:    resource.MustParse("500m"),
+			v1.ResourceMemory: resource.MustParse("550Mi"),
+		},
+	}
+}
+
+func DefaultBackendWorkerResourceRequirements() v1.ResourceRequirements {
+	return v1.ResourceRequirements{
+		Limits: v1.ResourceList{
+			v1.ResourceCPU:    resource.MustParse("1000m"),
+			v1.ResourceMemory: resource.MustParse("300Mi"),
+		},
+		Requests: v1.ResourceList{
+			v1.ResourceCPU:    resource.MustParse("150m"),
+			v1.ResourceMemory: resource.MustParse("50Mi"),
+		},
+	}
+}
+
+func DefaultCronResourceRequirements() v1.ResourceRequirements {
+	return v1.ResourceRequirements{
+		Limits: v1.ResourceList{
+			v1.ResourceCPU:    resource.MustParse("150m"),
+			v1.ResourceMemory: resource.MustParse("80Mi"),
+		},
+		Requests: v1.ResourceList{
+			v1.ResourceCPU:    resource.MustParse("50m"),
+			v1.ResourceMemory: resource.MustParse("40Mi"),
+		},
+	}
+}
+
+func DefaultSystemBackendUsername() string {
+	return "3scale_api_user"
+}
+
+func DefaultSystemBackendPassword() string {
+	return oprand.String(8)
```

Comment:
Could this potentially be "dangerous" for the stability of an existing deployment?
Each time that function is called would return a different password.
I think it should not be a problem right now because in the Operator options provider code is just used to be set as a default that is just used when the field does not already exist in the secret, but anyway I'd like to comment about this.

Reply:
I agree, can be changed

Reply:
To what would you change it?
From what I understand from other comments, the responsibility of instantiating the component itself(not the options) with the correct values for the options (being consistent between them and no erroneus default values) is of the options providers.
That's why I wrote that it should not be a problem right now because it's just used as the default, but at the same time is somewhat risky if we have a bug in the options providers code

Reply:
The behavior did not change respect to the current code base. On each reconciliation loop if a certain secret does not have some optional field, a entirely new default value will be generated. It should not be an issue because it is expected to be setting defaults only once.

So, how can be improved? having in mind that the operator only manages one 3scale instance, we could use process level (golang level) constants initialized at bootstrap time. This solutions would not work if the operator could manage several 3scale instances.

Reply:
:+1: yes that's why I said it should not be a problem right now because it's just used as the default, same as before. Both ways are risky.
I don't think it is a priority to change this compared to other things. I leave it up to you. To me changin it is not really a must at this moment.